### PR TITLE
Add logs to kubelet admit handle

### PR DESCRIPTION
**Pull request by tossmilestone**
Signed-off-by: Xiaoxi He <xxhe@alauda.io>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Since there are no logs when kubelet is executing admit handlers, it's diffcult to debug what's the reason of pod rejected just from the kubelet logs. The PR will add kubelet admit logs to help debugging with admit problems.

#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1942,7 +1943,18 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 	// TODO: out of resource eviction should have a pod admitter call-out
 	attrs := &lifecycle.PodAdmitAttributes{Pod: pod, OtherPods: pods}
 	for _, podAdmitHandler := range kl.admitHandlers {
+		if klog.V(4).Enabled() {
```
Comment (swatisehgal) [inline pkg/kubelet/kubelet.go:1946]: This check seems redundant here as we are using `klog.V(4).InfoS` later. Please refer to [this](https://github.com/kubernetes/klog/blob/main/klog.go#L1210).

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1942,7 +1943,18 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
...
+				"admitHandler", reflect.TypeOf(podAdmitHandler),
+				"pod", klog.KObj(pod),
+				"podUID", pod.UID)
+		}
 		if result := podAdmitHandler.Admit(attrs); !result.Admit {
+			klog.InfoS("Pod is rejected by AdmitHandler",
+				"admitHandler", reflect.TypeOf(podAdmitHandler),
+				"pod", klog.KObj(pod),
+				"podUID", pod.UID,
+				"result", result)
```
Comment (swatisehgal) [inline pkg/kubelet/kubelet.go:1957]: We should probably reconsider the use of log level 4 (debug log level) here and probably use log level 2 or log by default as these are useful log messages. Also, rather than logging for all admit handlers, I would recommend that you log once before returning from the function and follow the logging convention of logging messages in a single line:
```suggestion
	for _, podAdmitHandler := range kl.admitHandlers {
		if result := podAdmitHandler.Admit(attrs); !result.Admit {
		        klog.InfoS("Pod is rejected by AdmitHandler", "admitHandler", reflect.TypeOf(podAdmitHandler),"pod", klog.KObj(pod), "podUID", pod.UID,  "result", result)
			return false, result.Reason, result.Message
		}
	}
	
	kklog.InfoS("Pod Admitted", "pod", klog.KObj(pod), "podUID", pod.UID)
	return true, "", ""
```

**Review by swatisehgal — COMMENTED** [review]
Thanks for the PR, suggested a few changes. Also, we have a [`canRunPod` ](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1954)method   just below canAdmitPod method that has a very similar structure. Can we add similar log messages there as well?

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (tossmilestone) [reply pkg/kubelet/kubelet.go:1957]: Thanks for your suggestion!

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (tossmilestone) [reply pkg/kubelet/kubelet.go:1946]: @swatisehgal Because there is a `reflect.TypeOf` called in the log, so I use `klog.V(4).Enabled()` to reduce the performance impact on the log printing.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (tossmilestone) [reply pkg/kubelet/kubelet.go:1957]: @swatisehgal `kklog.InfoS("Pod Admitted", "pod", klog.KObj(pod), "podUID", pod.UID)` only prints the admitted pod, but have no handler name, it's also useful if we know the pod has been admitted by which admit handler.

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 3]
```diff
@@ -197,7 +197,6 @@ func (m *manager) RemoveContainer(containerID string) error {
 }
 
 func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
-	klog.InfoS("Topology Admit Handler")
```
[the remark this answers is not in the document]
Comment (tossmilestone) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:200]: Do you means to add a log to every `Admit` function?

**Comment by tossmilestone** [comment]
> Thanks for the PR, suggested a few changes. Also, we have a [`canRunPod` ](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1954)method just below canAdmitPod method that has a very similar structure. Can we add similar log messages there as well?

Yes, I will add the log to `canRunPod` as well

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tossmilestone) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:200]: @swatisehgal I have added an entry log to every admit handlers.

**Comment by tossmilestone** [comment]
@swatisehgal hi, thanks for you reply. I have done the changes, PTAL.

**Comment by tossmilestone** [comment]
@swatisehgal hi, the pr has been done for long days, please help to review it again, thanks!

File: pkg/kubelet/cm/container_manager_linux.go [hunk 4]
```diff
@@ -684,6 +684,9 @@ type resourceAllocator struct {
 }
 
 func (m *resourceAllocator) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	if attrs != nil && attrs.Pod != nil {
```
Comment (fromanirh) [inline pkg/kubelet/cm/container_manager_linux.go:687]: we inconditionally deference both `attrs` and `attrs.Pod` a couple lines below, so you can just do the same; if we want to be defensive, I think it's better done in a separate commit (or PR entirely)

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 5]
```diff
@@ -197,7 +197,9 @@ func (m *manager) RemoveContainer(containerID string) error {
...
+	if attrs != nil && attrs.Pod != nil {
```
Comment (fromanirh) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:200]: same

File: pkg/kubelet/eviction/eviction_manager.go [hunk 6]
```diff
@@ -134,6 +134,9 @@ func NewManager(
 
 // Admit rejects a pod if its not safe to admit for node stability.
 func (m *managerImpl) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	if attrs != nil && attrs.Pod != nil {
+		klog.InfoS("Eviction Admit Handler", "pod", klog.KObj(attrs.Pod), "podUID", attrs.Pod.UID)
+	}
```
Comment (fromanirh) [inline pkg/kubelet/eviction/eviction_manager.go:139]: (almost) the same

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -1966,10 +1967,12 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 	attrs := &lifecycle.PodAdmitAttributes{Pod: pod, OtherPods: pods}
 	for _, podAdmitHandler := range kl.admitHandlers {
 		if result := podAdmitHandler.Admit(attrs); !result.Admit {
+			klog.InfoS("Pod is rejected", "admitHandler", reflect.TypeOf(podAdmitHandler), "pod", klog.KObj(pod), "podUID", pod.UID, "result", result)
```
Comment (fromanirh) [inline pkg/kubelet/kubelet.go:1970]: I'm not completely sure we want to use `reflect` here, is it used elsewhere in the code/package?

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -1980,10 +1983,12 @@ func (kl *Kubelet) canRunPod(pod *v1.Pod) lifecycle.PodAdmitResult {
 
 	for _, handler := range kl.softAdmitHandlers {
 		if result := handler.Admit(attrs); !result.Admit {
+			klog.InfoS("Pod cannot run", "admitHandler", reflect.TypeOf(handler), "pod", klog.KObj(pod), "podUID", pod.UID, "result", result)
```
Comment (fromanirh) [inline pkg/kubelet/kubelet.go:1986]: same

File: pkg/kubelet/kubelet.go [hunk 9]
```diff
@@ -1966,10 +1967,12 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 	attrs := &lifecycle.PodAdmitAttributes{Pod: pod, OtherPods: pods}
 	for _, podAdmitHandler := range kl.admitHandlers {
 		if result := podAdmitHandler.Admit(attrs); !result.Admit {
+			klog.InfoS("Pod is rejected", "admitHandler", reflect.TypeOf(podAdmitHandler), "pod", klog.KObj(pod), "podUID", pod.UID, "result", result)
 			return false, result.Reason, result.Message
 		}
 	}
 
+	klog.InfoS("Pod Admitted", "pod", klog.KObj(pod), "podUID", pod.UID)
```
Comment (fromanirh) [inline pkg/kubelet/kubelet.go:1975]: admission rejection should be sporadic, but admission success should be way more common, so I think this should be `V(4)`

File: pkg/kubelet/lifecycle/handlers.go [hunk 10]
```diff
@@ -158,6 +158,9 @@ type appArmorAdmitHandler struct {
 }
 
 func (a *appArmorAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
+	if attrs != nil && attrs.Pod != nil {
+		klog.InfoS("AppArmor Admit Handler", "pod", klog.KObj(attrs.Pod), "podUID", attrs.Pod.UID)
+	}
```
Comment (fromanirh) [inline pkg/kubelet/lifecycle/handlers.go:163]: we inconditionally deference both `attrs` and `attrs.Pod` a couple lines below, so you can just do the same; if we want to be defensive, I think it's better done in a separate commit (or PR entirely)

File: pkg/kubelet/sysctl/allowlist.go [hunk 11]
```diff
@@ -110,6 +111,9 @@ func (w *patternAllowlist) validateSysctl(sysctl string, hostNet, hostIPC bool)
 // Admit checks that all sysctls given in pod's security context
 // are valid according to the allowlist.
 func (w *patternAllowlist) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	if attrs != nil && attrs.Pod != nil {
```
Comment (fromanirh) [inline pkg/kubelet/sysctl/allowlist.go:114]: we inconditionally deference both `attrs` and `attrs.Pod` a couple lines below, so you can just do the same; if we want to be defensive, I think it's better done in a separate commit (or PR entirely)

**Review by ffromani — COMMENTED** [review]
In general I'm in favour of this change, but we should not add these logs inconditionally.
Rejections should be logged at `V(3)` (or even `V(5)` IMHO), and accepts at `V(4)` (or `V(6)`)
